src/vit-testing/mainnet-lib/src/wallet/registration.rs

```diff
@@ -0,0 +1,190 @@
+use cardano_serialization_lib::chain_crypto::Blake2b256;
+use cardano_serialization_lib::crypto::{Ed25519Signature, PublicKey};
+use cardano_serialization_lib::metadata::{
+    GeneralTransactionMetadata, MetadataMap, TransactionMetadatum, TransactionMetadatumLabel,
+};
+use cardano_serialization_lib::utils::{BigNum, Int};
+use snapshot_lib::registration::Delegations;
+
+pub struct RegistrationBuilder<'a> {
+    wallet: &'a MainnetWallet,
+    delegations: Option<Delegations>,
+    slot_no: u64,
+}
+
+impl<'a> RegistrationBuilder<'a> {
+    pub fn new(wallet: &'a MainnetWallet) -> Self {
+        Self {
+            wallet,
+            delegations: None,
+            slot_no: 0,
+        }
+    }
+
+    pub fn to(mut self, delegations: Delegations) -> Self {
+        self.delegations = Some(delegations);
+        self
+    }
+
+    pub fn on(mut self, slot_no: u64) -> Self {
+        self.slot_no = slot_no;
+        self
+    }
+
+    pub fn build(self) -> GeneralTransactionMetadata {
+        let mut meta_map: MetadataMap = MetadataMap::new();
+
+        match self.delegations.expect("no registration target defined") {
+            Delegations::New(_delegations) => (),
+            Delegations::Legacy(legacy) => {
+                meta_map.insert(
+                    &TransactionMetadatum::new_int(&Int::new_i32(1)),
+                    &TransactionMetadatum::new_bytes(hex::decode(legacy.to_hex()).unwrap())
+                        .unwrap(),
+                );
+            }
+        };
+
+        meta_map.insert(
+            &TransactionMetadatum::new_int(&Int::new_i32(2)),
+            &TransactionMetadatum::new_bytes(self.wallet.stake_public_key().as_bytes()).unwrap(),
+        );
+        meta_map.insert(
+            &TransactionMetadatum::new_int(&Int::new_i32(3)),
+            &TransactionMetadatum::new_bytes(self.wallet.reward_address().to_address().to_bytes())
+                .unwrap(),
+        );
+        meta_map.insert(
+            &TransactionMetadatum::new_int(&Int::new_i32(4)),
+            &TransactionMetadatum::new_int(&Int::new(&BigNum::from(self.slot_no))),
+        );
+
+        let mut metadata = GeneralTransactionMetadata::new();
+        metadata.insert(
+            &BigNum::from(61284u32),
+            &TransactionMetadatum::new_map(&meta_map),
+        );
+
+        let meta_bytes = metadata.to_bytes();
+        let meta_bytes_hash = Blake2b256::new(&meta_bytes);
+        let signature = self.wallet.stake_key.sign(meta_bytes_hash.as_hash_bytes());
+
+        let mut meta_sign_map: MetadataMap = MetadataMap::new();
+
+        meta_sign_map.insert(
+            &TransactionMetadatum::new_int(&Int::new_i32(1)),
+            &TransactionMetadatum::new_bytes(signature.to_bytes()).unwrap(),
+        );
+
+        metadata.insert(
+            &BigNum::from(61285u32),
+            &TransactionMetadatum::new_map(&meta_sign_map),
+        );
+
+        metadata
+    }
+}
+
+pub trait GeneralTransactionMetadataInfo {
```

Comment:
I wonder about having a trait vs just a struct here. What's the advantage of this approach vs having a function `get_metadata_info(&self) -> GeneralTransactionMetadataInfo` on types that we need it for?

Reply:
I must confess that i introduced a trait for uncertain future because i feel like in other places in code (for example when testing registration transaction correctness etc.) would be handy. Currently yes, it looks like over engineering (however i think it encapsulated GeneralTrancationMetadata and registration builder doesn't know too much about it)

Reply:
My usual strategy would be to introduce the trait in the same PR that we introduce the second implementation (in case we never get round to it, or decide against it, etc).

But it's not a huge deal, happy to go with your judgement